### DISCUSSION TITLE
Adding computed values as grouping information with BIRT Designer 2022-03 (4.23)

### DISCUSSION THREAD

**Discussion by Markus-Elfring** [eclipse-birt/birt#975] [Q&A]
I put two fields into [a grouped listing report](https://eclipse.github.io/birt-website/docs/t_create-a-report#choose-a-template "Choosing a report template") of the tool “BIRT Designer 2022-03 (4.23)”.

* The first field is a name as the primary grouping criterium.
  I would appreciate if a running counter can be prepended to the group header display.
* The second field is an other identifier which should be treated as the detail row.
  I would like to add the number of items in the group to the footer row.

Which expressions should be specified for this purpose in the report layout? :thinking:

**Comment by Markus-Elfring** [comment]
:crystal_ball: Can the following example data help to clarify the construction of nice grouped listing reports for further software tests?

```CSV
name,id
f,1
f,2
g,3
g,4
g,5
```

**Comment by Markus-Elfring** [comment]
:thought_balloon: It seems that I am also looking for more current documentation according to the component “Aggregation Builder” of the tool “Eclipse BIRT Designer Version 4.9.0.v202203150031”.
I got the impression that further software extensions will become helpful. :thinking:

**Comment by hvbtup** [comment]
There is no need to code such a basic thing yourself.
- Create a data set that selects the rows
- Create a table
- Add a group to the table. Choose the column _name_ for the "Group on" field in the "Edit group" dialog.
- Now your table has an additional group header and group footer row.
- BIRT will also show the _name_ in a cell of the group header row. You might consider deleting the _name_ output from the _detail_ row.
- Drag and drop the "Aggregation" quick tool (Sigma/Summary-Symbol) from the palette to the _group footer row_.
- In the Aggregation Builder dialog, choose the COUNT function and the _id_ column the expression (choose from the drop-down-list). Don't forget to choose a meaningful "column binding name". Now your group footer shows the number of items for each _name_.
- The running count for the _name_ is more complicated, see below.

What you want is an output in the group header rows like

    1. f
    2. g

You can't just use the RUNNINGCOUNT function (try it out!, neither with "Aggrate On:" Table nor with "Aggregate On": Group xxx).
It will not contain the right values.
I'm not sure if RANK would help You can give it a try with "expression": _name_ and "Aggregate On": Table. 
With SQL, the correct analytical funtion would be DENSE_RANK.
If your dataset is SQL based, I recommend using SQL analytic functions.

It is certainly possible with BIRT, but I don't see an _obvious_ way.

**Reply by Markus-Elfring** [reply]
>It is certainly possible with BIRT, but I don't see an _obvious_ way.

I find this view interesting for further clarifications.


>I'm not sure if RANK would help You can give it a try with "expression": name and "Aggregate On": Table.

I do not see the expected counter value for the second name example after I constructed another dynamic text object as a header field accordingly.

**Comment by hvbtup** [comment]
Why do you create a dynamic text item at all?
Maybe you should attach the .rptdesign file here (wrapped in a ZIP archive).

**Reply by Markus-Elfring** [reply]
>Why do you create a dynamic text item at all?

I tried the following code out for the construction of a potentially desirable text as a header field according to expression building.
```
row["rank"] + ". " + row["name"]
```

**Reply by hvbtup** [reply]
Maybe you should attach the .rptdesign file here (wrapped in a ZIP archive).

**Reply by Markus-Elfring** [reply]
Would you suggest to combine the computed value (aggregation) with the field content by any other means than the dialogue “Expression Builder (JavaScript)”?
[Test_data5-20220523.rptdesign.zip](https://github.com/eclipse/birt/files/8763424/Test_data5-20220523.rptdesign.zip)

**Comment by hvbtup** [comment]
I suggested this to keep things simple at first and to avoid pitfalls like what happens here.

Your expression is 
![grafik](https://user-images.githubusercontent.com/7644776/170062605-e50d9d8e-265c-40bb-a80e-e0b56a7f8cfd.png)

And the binding is
![grafik](https://user-images.githubusercontent.com/7644776/170062761-af707bfb-dfed-4bdc-b45b-43508d9c6026.png)

Note that _rank_ is a number, the rest of the expression is text. **The operator `+` will try to convert the text `'.'` to a number and this will fail of course.**

If you choose "Run"/"View Report"/"As HTML" you will probably see red error messages. Not sure, because I can't run your report (the flat file with the data is missing).

Even seemingly simple tasks can result in hard-to-detect errors with Javascript. That's why I personally dislike Javascript.

You can convert a number to a text if you add the number to an empty string. So, 

    "" + row["rank"] + ". " + row["name"]

will solve this problem.

**Reply by Markus-Elfring** [reply]
>Note that rank is a number, the rest of the expression is text.

This view is generally fine.


>**The operator + will try to convert the text '.' to a number** …

I observed other software behaviour.

Two characters (and a name for this test case) should be appended to the string representation of an integral value, shouldn't they?

The text display “`2. g`” should be generated somehow for one report group header here.

**Comment by hvbtup** [comment]
Implicit type conversion is causing sooo many errors. For professional programming, I would prefer a language that does not allow implicit conversion at all (not even between different number data types). I think such languages exist, but are rather exotic.

**Comment by hvbtup** [comment]
From your not-so-precise description of the behavior I concluded that the report failed to run.
Please, add some screenshots to your descriptions next time.

Your report works for me, after I took the time to create a CSV file myself.
It's just that RANK returns 3 for the second group, which is what I expected. The RANK function behaves like the RANK SQL function. The SQL DENSE_RANK function would return 2 for the second group, but BIRT doesn't have such an aggregate function.

**Reply by Markus-Elfring** [reply]
>From your not-so-precise description of the behavior

I tend to increase corresponding details on demand.


>I concluded that the report failed to run.

I suggest to reconsider such conclusions a bit more.


>Your report works for me, after I took the time to create a CSV file myself.

Thanks for your positive feedback.


>It's just that RANK returns 3 for the second group,

I am observing the same test result.


>which is what I expected.

Our expectations are evolving also according to desirable software behaviour, aren't they?


>The RANK function behaves like the RANK SQL function.

Will any descriptions become more helpful also for the graphical user interface of the report designing tool?


>The SQL DENSE_RANK function would return 2 for the second group,

Such a result would be nice for my report approach.


>but BIRT doesn't have such an aggregate function.

* Would any software users like to share more experiences around this functionality?
* How will the chances evolve to extend the supported aggregation possibilities?

**Comment by hvbtup** [comment]
[DENSE_RANK.zip](https://github.com/eclipse/birt/files/8764262/DENSE_RANK.zip)

This possible solution requires a little scripting.
What I added to the report:
- A report variable
- some code in the beforeOpen and the onFetch event of the DataSet to initialize and update the variable
- Displaying the value of the report variable in another group header cell. Note the values 1 and 2 for the first and second group.
![grafik](https://user-images.githubusercontent.com/7644776/170078252-d36b8fcc-d64c-4047-96b0-1594e394c19b.png)

**Reply by Markus-Elfring** [reply]
>This possible solution requires a little scripting.

Thanks for your demonstration of another software adjustment.
Will the chances grow then to improve the documentation accordingly?

**Comment by hvbtup** [comment]
> Will the chances grow then to improve the documentation accordingly?

Chances will grow if _you_ and others help.
For example, you can help improving the web site (see #893).

You can also help by providing features that you are missing. I created issue #981 for the missing DENSE_RANK aggregate function. 
Feel free to create a PR to implement this (see https://eclipse.github.io/birt-website/docs/how2contribute).

> I tend to increase corresponding details on demand.

You should aim to provide a clear and precise description of the issue to help other peaople to understand the issue right from the start.
If possible, you should also add an example report including all necessary data (ideally using the ClassicModels builtin example data source).
